### PR TITLE
CTC Loss Operator

### PR DESCRIPTION
**Pull request by sbodenstein**
This PR merges WarpCTC (https://github.com/baidu-research/warp-ctc) into MXNet, so that MXNet has a native CTC loss layer. Some questions/issues:

- This currently depends on ModernGPU, which is a header only GPU library. This dependency should eventually be removed in favor of CUB. 
- Lint: I haven't conformed ModernGPU and WarpCTC to MXNet code style. Should this be done, or is there a way to set the linter to ignore these folders?
- I don't have any license info in the WarpCTC files. What should I put?
- I have already removed dependency on some custom reduction routines, and am using MShadow instead. So the GPU timings might be a bit different from WarpCTC.

Some design issues:
- This layer uses the following sizes for inputs and outputs: data has dims `(sequence_length, batch_size, alphabet_size + 1)`, labels has dims `(batch_size, label_sequence_length)`, and output `(batch_size)`. The output is a loss for every element in the batch.
- The implementation supports variable length label sequences via 0-padding. A separate PR will implement variable-length sequence support for the data input.

@piiswrong

### CODE REVIEW

**Comment by piiswrong** [comment]
@ap-hynninen @ptrendx Do you know if it's easy to replace mordern gpu with CUB?

**Comment by ap-hynninen** [comment]
At quick glance, warp-ctc is only using scan and mergesort algorithms. Scan is already in CUB, mergesort is not. So most of the work would be to switch to radix sort. Shouldn't be too much work but does require understanding of what the code is doing.

**Comment by sbodenstein** [comment]
@piiswrong: what strategy should we use for this then? Merge this, and then remove the ModernGPU dependency later?

**Comment by piiswrong** [comment]
Sounds good to me

**Comment by sbodenstein** [comment]
Ok, great! Then what about the questions I had (for example, dealing with the Linter, and crediting WarpCTC people?)

**Comment by piiswrong** [comment]
A license should be included in ctc_inlude folder.

As for lint, we can disable it for those folders. But I don't know how...

**Comment by sbodenstein** [comment]
@piiswrong: one option is to add an excluded folder option to https://github.com/dmlc/dmlc-core/blob/master/scripts/lint.py#L145, and then modify the MXNet `Makefile`.

**Comment by piiswrong** [comment]
sounds good

**Comment by sbodenstein** [comment]
Apologies, have been away at a conference for the past week, and haven't had time to follow up on the changes needed to DML-CORE. I will either do it today, or next week Tuesday. Unless someone else wants to do it first.

**Comment by sbodenstein** [comment]
@piiswrong: this is building correctly on both GPU + CPU, but its failing on the C++ package generation. I don't know much about this, could you help me? Lots of errors are generated from Jenkins, a small sample:
```
cpp-package/include/mxnet-cpp/op.h: In function 'mxnet::cpp::Symbol mxnet::cpp::CTCLoss(const string&, int, int)':
cpp-package/include/mxnet-cpp/op.h:61:30: error: 'data' was not declared in this scope
            .SetParam("data", data)

File: src/operator/contrib/ctc_loss.cc [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+applies a softmax to each vector, which then becomes a vector of probabilities 
+over the alphabet. Note that the 0th element of this vector is reserved for the 
+special blank character.
+
+See *Connectionist Temporal Classification: Labelling Unsegmented 
+Sequence Data with Recurrent Neural Networks*, A. Graves *et al*. for more 
+information.
+
+)code" ADD_FILELINE)
+    .add_argument("data", "ndarray-or-symbol", "Input data to the ctc_loss op.")
```
Comment (piiswrong) [inline src/operator/contrib/ctc_loss.cc]: change to NDArray-or-Symbol
should solve the cpp package problem

**Comment by piiswrong** [comment]
btw why labels instead of label?

File: src/operator/contrib/ctc_loss.cc (hunk 1 above)
Comment (sbodenstein) [inline src/operator/contrib/ctc_loss.cc]: ah, thanks!

**Comment by sbodenstein** [comment]
@piiswrong: there are multiple labels. But I guess CrossEntropyLoss can also have multiple labels, and this is called "label". Should I rename it?

**Comment by piiswrong** [comment]
yeah I think label is more consistent

**Comment by sbodenstein** [comment]
@piiswrong: all checks are passing, this seems ready for merging.

**Comment by piiswrong** [comment]
looks good.
One question: is it possible to respect out_grad so that we can do something like
loss = ctx_loss(data, label)
loss = loss * per_sample_weight
to weight each instance?

**Comment by sbodenstein** [comment]
@piiswrong: yes, all loss functions we use in Mathematica must respect this, so I did add handling for this. See [this](https://github.com/dmlc/mxnet/pull/5834/files#diff-5574eb3b2aa10440c134dbe9c1599dc2R206), and check if its correct.

**Comment by piiswrong** [comment]
cool

**Comment by sbodenstein** [comment]
I do want to add support for variable length inputs (it currently only supports variable length labels). But I will do that in a separate PR, so this can be merged now.

File: src/operator/contrib/ctc_loss.cc [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  std::vector<TShape> out_shape, aux_shape;
+  std::vector<int> out_type, aux_type;
+  CHECK(InferType(in_type, &out_type, &aux_type));
+  CHECK(InferShape(in_shape, &out_shape, &aux_shape));
+  DO_BIND_DISPATCH(CreateOp, param_, (*in_type)[0]);
+}
+
+DMLC_REGISTER_PARAMETER(CTCLossParam);
+
+MXNET_REGISTER_OP_PROPERTY(CTCLoss, CTCLossProp)
```
Comment (piiswrong) [inline src/operator/contrib/ctc_loss.cc]: @sbodenstein I over looked this. This should be contrib.ctc_loss since its experimental.
